runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java

```diff
@@ -645,6 +655,11 @@ private void declareForEach(ExtensionDeclarer extensionDeclarer, ClassTypeLoader
     forEach.withOutput().ofType(VOID_TYPE);
     forEach.withOutputAttributes().ofType(VOID_TYPE);
 
+    configurerFactory.create()
+        .setKeysResolver(new NullMetadataResolver(), "collection", STRING_TYPE, false)
```

Comment:
not a string

Reply:
Unfortunately we can't make it be the same type of the parameter in this case, for a couple of reasons:
1- The value is expected to be an expression. We can't resolve expressions when doing type resolution because we don't have an execution context. So the only value we can provide is the unresolved expression. The closest metadata type we have to represent that is `String`. If we were to use the parameter type here (`Iterable<Object>`), there would be a mismatch. Fortunately, `String` is also the key type that the `InpuTypeResolver` needs in this case.
2- There is [a check](https://github.com/mulesoft/mule/blob/master/modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java#L303) in the `MetadataKeyIdObjectResolver` which explicitly says `'Iterable' type is invalid for MetadataKeyId parameters, use String type instead`. It seems only Boolean, String or Object metadata types are allowed.

Reply:
please explain this is a comment

Reply:
The type there is used as the expected type of the metadata key object (see [MetadataKeyIdObjectResolver](https://github.com/mulesoft/mule/blob/master/modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/metadata/MetadataKeyIdObjectResolver.java)). Up to this point it has always been the case that the type of the metadata key object is the same of the parameter annotated with `MetadataKeyId`.

This case is different, the [parameter is of ArrayType](https://github.com/mulesoft/mule/blob/master/runtime-extension-model/src/main/java/org/mule/runtime/core/api/extension/provider/MuleExtensionModelDeclarer.java#L617) but our metadata key (which is derived from it) is **the expression** itself. 

If we were to say that the type of the metadata key needs to be an `ArrayType` just like the parameter, not only we don't have a way to extract an array from the expression (remember we can't evaluate it at design time), but also it wouldn't be useful for our `InputTypeResolver` that depends on that key.

Our `InputTypeResolver` needs the expression so that it can give it to the `ExpressionLanguageMetadataService` so that it can tell it what would be the output type of the expression considering other types bound in context (payload, attributes, vars).

Because (AFAIK) we don't have a special MetadataType to represent an expression, I'm using StringType.